### PR TITLE
Surfrad data reader

### LINKED ISSUE #590: add SURFRAD data reader
Add data reader for the NOAA SURFRAD network.

https://www.esrl.noaa.gov/gmd/grad/surfrad/ 

This should be a new module under the pvlib.iotools subpackage proposed in #272.

### PR DESCRIPTION
**Pull request by lboeman**
pvlib python pull request guidelines
====================================

Thank you for your contribution to pvlib python! You may delete all of these instructions except for the list below.

You may submit a pull request with your code at any stage of completion.

The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items below:

 - [x] Closes #590
 - [x] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [x] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [x] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [x] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [x] Code quality and style is sufficient. Passes LGTM and SticklerCI checks.
 - [x] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.

Brief description of the problem and proposed solution (if not already fully described in the issue linked to above):

### CODE REVIEW

File: pvlib/test/test_surfrad.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from pvlib.iotools import surfrad
+
+test_dir = os.path.dirname(
+    os.path.abspath(inspect.getfile(inspect.currentframe())))
+testfile = os.path.join(test_dir, '../data/surfrad-slv16001.dat')
+network_testfile = ('ftp://aftp.cmdl.noaa.gov/data/radiation/surfrad/'
+                    'Alamosa_CO/2016/slv16001.dat')
+
+
+def test_read_surfrad():
```
Comment (wholmgren) [inline pvlib/test/test_surfrad.py]: I don't think this test is doing anything for us so we can delete it

Comment (wholmgren) [reply pvlib/test/test_surfrad.py]: sorry for setting a bad example in `test_tmy.py`

File: pvlib/test/test_surfrad.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+def test_read_surfrad_network():
+    surfrad.read_surfrad(network_testfile)
+
+
+def test_read_surfrad_columns_exist():
+    data, _ = surfrad.read_surfrad(testfile)
+    assert 'zen' in data.columns
+    assert 'temp' in data.columns
+    assert 'par' in data.columns
+    assert 'pressure' in data.columns
```
Comment (wholmgren) [inline pvlib/test/test_surfrad.py]: this function will need to change if we add a variable mapping option

File: pvlib/test/test_surfrad.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def test_format_index():
+    start = Timestamp('20160101 00:00')
+    expected_index = DatetimeIndex(start=start, periods=1440, freq='1min')
```
Comment (wholmgren) [inline pvlib/test/test_surfrad.py]: I think you can supply `tz='UTC'` here and skip the localize call below. Seems a little more straightforward to me. I had to look up if localize modifies in place or not.

File: pvlib/test/test_surfrad.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    expected_index.tz_localize('UTC')
+    actual, _ = surfrad.read_surfrad(testfile)
+    assert actual.index.equals(expected_index)
+
+
+def test_read_surfrad_metadata():
+    expected = {'name': 'Alamosa',
+                'latitude': '37.70',
```
Comment (wholmgren) [inline pvlib/test/test_surfrad.py]: the values here and below should be numeric. probably all floats.

File: pvlib/test/test_surfrad.py (hunk 3 above)
Comment (lboeman) [reply pvlib/test/test_surfrad.py]: Good catch, thanks. This was a naive "fix" when I was reading the test failure incorrectly, I wasn't setting the timezone correctly in read_surfrad.
It looks like pandas 0.14.0 has some issues with setting the timezone the way you proposed though, so i'll have to figure that out.

**Comment by lboeman** [comment]
@wholmgren Looks like the `tz` parameter is added to pandas.DatetimeIndex in 0.15.0

**Comment by wholmgren** [comment]
Let's bump the minimum requirements to 0.15.0.

@mikofski will also appreciate the `tz_localize(None)` feature.

http://pandas.pydata.org/pandas-docs/stable/whatsnew.html#timezone-handling-improvements

**Comment by lboeman** [comment]
The bump to pandas 0.15.0 is causing other tests to fail and after testing it looks like all versions from 15-18.1 all produce errors in the tests. So it's probably best to roll this back to 0.14.0?

**Comment by wholmgren** [comment]
The min failures are due to a couple of issues. 

`test_clearsky` and `test_solarposition` failures are solved with code suggested in #598. That should be addressed in a different PR, possibly #583. 

`test_pvsystem` failures are due to https://github.com/pandas-dev/pandas/issues/13006 and can be avoided by changing pvsystem.py lines 2651-2652 to:

```
    v_lim_upper = float(np.nanmax([v_max, vdc_max, mppt_hi]) * (1 + vtol))
    v_lim_lower = float(np.nanmax([v_min, mppt_low]) * (1 - vtol))
```

**Comment by wholmgren** [comment]
@lboeman can you merge the latest pvlib-python/master into your branch and make the changes to pvsystem.py proposed above? I think that should make the tests pass.

**Comment by lboeman** [comment]
@wholmgren Updated those lines and all tests are passing now.

**Comment by cwhanse** [comment]
Just a thought, not requesting any change: the `format_index` function seems likely to be useful in other contexts.  If I had an idea how to name it so that someone else would recognize its purpose from the name, I would suggest moving it to `it.tools` or something similar.

**Comment by wholmgren** [comment]
@cwhanse do you know of other data sources that use the SURFRAD time format? If not, I suggest we leave it as for now.

File: docs/sphinx/source/whatsnew/v0.6.1.rst [hunk 5]
```diff
@@ -30,6 +32,11 @@ Enhancements
   :py:func:`pvlib.iotools.read_srml_month_from_solardat` to read University of
   Oregon Solar Radiation Monitoring Laboratory data. (:issue:`589`)
 
+* :func:`~pvlib.solarposition.rise_set_transit_ephem` returns sunrise, sunset and transit times using pyephem
+* Created :py:func:`pvlib.iotools.read_srml` and :py:func:`pvlib.iotools.read_srml_month_from_solardat`
+  to read University of Oregon Solar Radiation Monitoring Laboratory data. (:issue:`589`)
+  
+* Created :py:func:`pvlib.iotools.read_surfrad` to read NOAA SURFRAD data. (:issue:`590`)
```
Comment (wholmgren) [inline docs/sphinx/source/whatsnew/v0.6.1.rst]: @lboeman your rebase introduced duplicate entries here

Comment (lboeman) [reply docs/sphinx/source/whatsnew/v0.6.1.rst]: Thanks! Fixed.

**Pull request merged**